calyx-py/test/correctness/queues/rr_queues/rr_queue.py

```diff
@@ -0,0 +1,222 @@
+
+def invoke_subqueue(queue_cell, cmd, value, ans, err) -> cb.invoke:
+   """Invokes the cell {queue_cell} with:
+   {cmd} passed by value
+   {value} passed by value
+   {ans} passed by reference
+   {err} passed by reference
+   """
+   return cb.invoke(
+       queue_cell,
+       in_cmd=cmd,
+       in_value=value,
+       ref_ans=ans,
+       ref_err=err,
+   )
+
+def insert_pifo(
+   prog,
+   name,
+   fifos,
+   boundaries,
+   n_flows, # the number of flows
+   queue_len_factor=QUEUE_LEN_FACTOR,
+   stats=None,
+   static=False,
+):
+   """Inserts the component `pifo` into the program.
+   """
+
+   pifo: cb.ComponentBuilder = prog.component(name)
+   cmd = pifo.input("cmd", 2) # the size in bits is 2
+   # If this is 0, we pop. If it is 1, we peek.
+   # If it is 2, we push `value` to the queue.
+   value = pifo.input("value", 32)  # The value to push to the queue
+
+   fifo_cells = []
+   for num in range(len(fifos)):
+       name = "queue_" + str(num)
+       cell = pifo.cell(name, fifos[num])
+       fifo_cells.append(cell)
+
+   # If a stats component was provided, declare it as a cell of this component.
+   if stats:
+       stats = pifo.cell("stats", stats, is_ref=True)
+
+   flow = pifo.reg(32, "flow")  # The flow to push to: 0 to n.
+   # We will infer this using a separate component;
+   # it is a function of the value being pushed.
+  
+   ans = pifo.reg(32, "ans", is_ref=True)
+   # If the user wants to pop, we will write the popped value to `ans`.
+
+   err = pifo.reg(1, "err", is_ref=True)
+   # We'll raise this as a general error flag for overflow and underflow.
+
+   length = pifo.reg(32, "length")  # The active length of the PIFO.
+
+   # A register that marks the next sub-queue to `pop` from.
+   hot = pifo.reg(32, "hot")
+   og_hot = pifo.reg(32, "og_hot")
+   copy_hot = pifo.reg_store(og_hot, hot.out) # og_hot := hot.out
+   restore_hot = pifo.reg_store(hot, og_hot.out) # hot := og_hot.out
+
+   max_queue_len = 2**queue_len_factor
+
+   adder = pifo.add(32, "adder_reg")
+   div_reg = pifo.div_pipe(32, "div_reg")
+   i = pifo.reg(32, "i")
+
+   # Some equality checks.
+   hot_eq_n = pifo.eq_use(hot.out, n_flows-1, cellname="hot_eq_n") #bc 0-based indexing
+   len_eq_0 = pifo.eq_use(length.out, 0, cellname="len_eq_0")
+   len_eq_max_queue_len = pifo.eq_use(length.out, max_queue_len, cellname="len_eq_maxq")
+   cmd_eq_0 = pifo.eq_use(cmd, 0, cellname="cmd_eq_0")
+   cmd_eq_1 = pifo.eq_use(cmd, 1, cellname="cmd_eq_1")
+   cmd_eq_2 = pifo.eq_use(cmd, 2, cellname="cmd_eq_2")
+   err_eq_0 = pifo.eq_use(err.out, 0, cellname="err_eq_0")
+   err_neq_0 = pifo.neq_use(err.out, 0, cellname="err_neq_0")
+
+   incr_hot = pifo.incr(hot)
+   raise_err = pifo.reg_store(err, 1, "raise_err")  # err := 1
+   lower_err = pifo.reg_store(err, 0, "lower_err")  # err := 0
+   reset_hot = pifo.reg_store(hot, 0, "reset_hot") # hot := 0
+
+   len_incr = pifo.incr(length)  # len++
+   len_decr = pifo.decr(length)  # len--
+
+   handles = []
```

Comment:
This is a really cool list that does a lot, I think we need to (1) give it a more descriptive name, and (2) have a comment explaining what is going on! I know the answer from our conversations, but here's your chance to explain the cleverness to all!